### PR TITLE
Use faster iterative algorithm to compute dominators for small CFGs

### PR DESCRIPTION
**Pull request by ddegazio**
#### 64f40e1806635d78dfe9a758585b5598d8793034
<pre>
Use faster iterative algorithm to compute dominators for small CFGs
<a href="https://bugs.webkit.org/show_bug.cgi?id=276977">https://bugs.webkit.org/show_bug.cgi?id=276977</a>
<a href="https://rdar.apple.com/problem/132363948">rdar://problem/132363948</a>

Reviewed by Yusuke Suzuki.

Implements the dominance algorithm described in &quot;A Simple, Fast Dominance Algorithm&quot;
(Cooper, Harvey, Kennedy 2001), and uses it over Lengauer-Tarjan when computing
dominators for graphs smaller than 20,000 nodes. On the JetStream 2 benchmark, this
means we compute dominance about 60% faster than Lengauer-Tarjan, although this
doesn&apos;t seem to translate to a measurable progression overall.

* Source/WTF/wtf/Dominators.h:
(WTF::Dominators::Dominators):
(WTF::Dominators::IterativeDominance::IterativeDominance):
(WTF::Dominators::IterativeDominance::computeReversePostorder):
(WTF::Dominators::IterativeDominance::intersect):
(WTF::Dominators::IterativeDominance::compute):
(WTF::Dominators::IterativeDominance::immediateDominator):

Canonical link: <a href="https://commits.webkit.org/281359@main">https://commits.webkit.org/281359@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/04ead2c7693a1baae286aac74b9fb9463d7b206f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63425 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10023 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48299 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7032 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61540 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51514 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29129 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32987 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8769 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8957 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52602 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54938 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9047 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65157 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58755 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3438 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8947 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55643 "Found 1 new test failure: media/media-source/media-source-abort-resets-parser.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3449 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51509 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55755 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2860 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80512 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8908 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34669 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13957 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35752 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36838 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35497 "Built successfully") | | |

### CODE REVIEW

File: Source/WTF/wtf/Dominators.h [hunk 1]
```diff
@@ -25,9 +25,11 @@
 
 #pragma once
 
+#include <wtf/BitSet.h>
 #include <wtf/CommaPrinter.h>
 #include <wtf/FastBitVector.h>
 #include <wtf/GraphNodeWorklist.h>
+#include <wtf/MonotonicTime.h>
```
Comment (hyjorc1) [inline Source/WTF/wtf/Dominators.h]: Do we need this?

File: Source/WTF/wtf/Dominators.h [hunk 2]
```diff
@@ -294,6 +314,129 @@ class Dominators {
...
+            RELEASE_ASSERT(graph.numNodes() < std::numeric_limits<int16_t>::max());
+            m_idoms.fill(undefinedIdom, graph.numNodes());
+        }
+
+        void computeReversePostorder()
+        {
+            BitSet<std::numeric_limits<int16_t>::max()> visited;
+            visited.clearAll();
+
+            Vector<int16_t> workList;
```
Comment (Constellation) [inline Source/WTF/wtf/Dominators.h]: Should we have some inlineCapacity here?

File: Source/WTF/wtf/Dominators.h [hunk 3]
```diff
@@ -294,6 +314,129 @@ class Dominators {
...
+
+            while (workList.size()) {
+                int16_t index = workList.takeLast();
+                if (index < 0) {
+                    // Negative indices mark nodes we're revisiting.
+                    m_order.append(~index);
+                    continue;
+                }
+                const auto& successors = m_graph.successors(m_graph.node(index));
+                if (!successors.size()) {
```
Comment (Constellation) [inline Source/WTF/wtf/Dominators.h:360]: `successors.isEmpty()`

File: Source/WTF/wtf/Dominators.h [hunk 4]
```diff
@@ -294,6 +314,129 @@ class Dominators {
...
+        {
+            if (block == m_graph.root())
+                return nullptr;
+            return m_graph.node(m_idoms[m_graph.index(block)]);
+        }
+
+    private:
+        Graph& m_graph;
+        Vector<uint16_t, 64> m_idoms;
+        Vector<uint16_t, 64> m_order;
```
Comment (Constellation) [inline Source/WTF/wtf/Dominators.h]: Maybe, naming `m_reversePostorderedNodes`?

File: Source/WTF/wtf/Dominators.h [hunk 5]
```diff
@@ -294,6 +314,129 @@ class Dominators {
...
+        void compute()
+        {
+            computeReversePostorder();
+
+            bool fixpoint = false;
+            int16_t rootIndex = m_graph.index(m_graph.root());
+            m_idoms[rootIndex] = rootIndex;
+            ASSERT(m_order[0] == rootIndex);
+            while (!fixpoint) {
+                fixpoint = true;
```
Comment (Constellation) [inline Source/WTF/wtf/Dominators.h]: We usually use `changed` boolean for fixpoint computation.

**Review by Constellation — APPROVED** [review]
r=me

File: Source/WTF/wtf/Dominators.h (hunk 1 above)
Comment (ddegazio) [reply Source/WTF/wtf/Dominators.h]: Ah, no, good catch - that was just for measuring the time we spend computing dominators. I'll remove it.

File: Source/WTF/wtf/Dominators.h (hunk 2 above)
Comment (ddegazio) [reply Source/WTF/wtf/Dominators.h]: Almost definitely yes, I'll add one!

File: Source/WTF/wtf/Dominators.h (hunk 3 above)
Comment (ddegazio) [reply Source/WTF/wtf/Dominators.h:360]: Actually I don't think that's available - `successors` is not a `Vector` but a `SuccessorList`, which is not assumed to have an `isEmpty()` method. So I think we have to check the `size()` here - it's also what the existing dominance algorithm does.

File: Source/WTF/wtf/Dominators.h (hunk 5 above)
Comment (ddegazio) [reply Source/WTF/wtf/Dominators.h]: Sure, I can rename it.

File: Source/WTF/wtf/Dominators.h (hunk 4 above)
Comment (ddegazio) [reply Source/WTF/wtf/Dominators.h]: Seems reasonable to me.

**Pull request merged**